src/Apps/Legion/Components/CookSource/CookSource.tsx

```diff
@@ -0,0 +1,284 @@
+    ITable
+} from '../../Adapters/Standalone/DataManagement/Models/DataManagementAdapter.types';
+import { useADXAdapter } from '../../Hooks/useADXAdapter';
+import {
+    IADXAdapterTargetContext,
+    IADXConnection,
+    ICookSource,
+    IPIDDocument
+} from '../../Models';
+import produce from 'immer';
+import { WritableDraft } from 'immer/dist/internal';
+import { getTableSchemaTypeFromTable } from '../../Services/DataPusherUtils';
+import { AppDataContext } from '../../Contexts/AppDataContext/AppDataContext';
+
+const debugLogging = false;
+const logDebugConsole = getDebugLogger('CookSource', debugLogging);
+
+const defaultTimeSeriesSource: ICookSource = {
+    cluster: null,
+    database: null,
+    table: null,
+    twinIdColumn: null,
+    tableType: null
+};
+
+const defaultPIDSource: ICookSource = {
+    pidUrl: ''
+};
+
+const CookSource: React.FC<ICookSourceProps> = (props) => {
+    const {
+        onSourceTypeChange,
+        onSourceChange,
+        onGetTableData,
+        targetAdapterContext = AppDataContext,
+        isClusterVisible = true
+    } = props;
+
+    // state
+    const [selectedSourceType, setSelectedSourceType] = useState<SourceType>(
+        SourceType.Timeseries
+    );
+    const [selectedSource, setSelectedSource] = useState<ICookSource>(
+        defaultTimeSeriesSource
+    );
+    const [sourceTableColumnOptions, setSourceTableColumnOptions] = useState<
+        Array<IDropdownOption>
+    >([]);
+
+    // hooks
+    const { t } = useTranslation();
+    const adapter = useADXAdapter(
+        targetAdapterContext as React.Context<IADXAdapterTargetContext>
+    );
+
+    const getTableState = useAdapter({
+        adapterMethod: (params: IGetTableAdapterParams) =>
+            adapter.getTable(
+                params.databaseName,
+                params.tableName,
+                TIMESTAMP_COLUMN_NAME
+            ),
+        isAdapterCalledOnMount: false,
+        refetchDependencies: [adapter]
+    });
+
+    // callbacks
+    const handleSourceTypeChange = useCallback(
+        (_event, option: IDropdownOption) => {
+            setSelectedSourceType(option.key as SourceType);
+            onSourceTypeChange?.(option.key as SourceType);
+            const newSource =
+                option.key === SourceType.Timeseries
+                    ? defaultTimeSeriesSource
+                    : defaultPIDSource;
+            setSelectedSource(newSource);
+            onSourceChange?.(newSource);
+        },
+        [onSourceChange, onSourceTypeChange]
+    );
+    const handleSourceClusterChange = useCallback(
+        (clusterUrl: string) => {
+            const newSource: IADXConnection = {
+                ...(selectedSource as IADXConnection),
+                cluster: clusterUrl,
+                database: null,
+                table: null,
+                twinIdColumn: null,
+                tableData: null
+            };
+            setSelectedSource(newSource);
```

Comment:
seeing so many operations on different parts of state makes me think maybe we should be using a reducer here.

Reply:
There is only 3 state variables, will keep it as is now.